src/Microsoft.Health.Dicom.SqlServer/Features/Common/SqlStoreFactory.cs

```diff
@@ -0,0 +1,39 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Linq;
+using EnsureThat;
+using Microsoft.Health.Dicom.Core.Features.Common;
+using Microsoft.Health.Dicom.SqlServer.Feature.Common;
+using Microsoft.Health.SqlServer.Features.Schema;
+
+namespace Microsoft.Health.Dicom.SqlServer.Features.ExtendedQueryTag
+{
+    /// <summary>
+    ///  Sql implementation of IStoreFactory, which takes multiple versioned stores, and return the one that should be used.
+    /// </summary>
+    /// <typeparam name="TVersionedStore">Type of versioned store.</typeparam>
+    /// <typeparam name="TStore">Type of the store.</typeparam>
+    internal class SqlStoreFactory<TVersionedStore, TStore> : IStoreFactory<TStore>
+        where TVersionedStore : IVersioned, TStore
+    {
+        private readonly SchemaInformation _schemaInformation;
+        private readonly IEnumerable<TVersionedStore> _versionedStores;
```

Comment:
Should this be an `IEnumerable<T>` or should this instead be a dictionary keyed on version?

Reply:
1. Regarding `IEnumerable<T>`.  The TVersionedStore need to implement both IVersioned and TStore, former is required for pickup correct version based on current schema version, latter is the actual store consumed by services.
2. Regarding dictionary. I'm not sure if direct inject could handle that...

Reply:
I meant within this class simply load the stores into a dictionary instead of searching the enumerable for a version. However, perhaps there is a way in the DI API to return the proper `TVersionedStore` based on the `SchemaInformation`'s Current version. However, if there is such an API, the usage of such a service would need to be scoped so it can re-resolve over time.

Reply:
I think dictionary has better performance as we have more and more versioned stores.